lib/project-folder-dnd-handler.coffee

```diff
@@ -0,0 +1,168 @@
+    element[0]?.updateTooltip()
+    @removeDropTargetClasses()
+    @removePlaceholder()
+
+  onDrop: (event) =>
+    event.preventDefault()
+    {dataTransfer} = event.originalEvent
+
+    # TODO: support dragging folders from the filesystem -- electron needs to add support first
+    return unless dataTransfer.getData('atom-event') is 'true'
+
+    fromWindowId = parseInt(dataTransfer.getData('from-window-id'))
+    fromRootPath  = dataTransfer.getData('from-root-path')
+    fromIndex     = parseInt(dataTransfer.getData('project-root-index'))
+    fromRootIndex = parseInt(dataTransfer.getData('from-root-index'))
+
+    toIndex = @getDropTargetIndex(event)
+
+    @clearDropTarget()
+
+    if fromWindowId is @getWindowId()
+      unless fromIndex is toIndex
+        projectPaths = atom.project.getPaths()
+        projectPaths.splice(fromIndex, 1)
+        if toIndex > fromIndex then toIndex -= 1
+        projectPaths.splice(toIndex, 0, fromRootPath)
+        atom.project.setPaths(projectPaths)
+    else
+      projectPaths = atom.project.getPaths()
+      projectPaths.splice(toIndex, 0, fromRootPath)
+      atom.project.setPaths(projectPaths)
+
+      if not isNaN(fromWindowId)
+        # Let the window where the drag started know that the tab was dropped
+        browserWindow = @browserWindowForId(fromWindowId)
+        browserWindow?.webContents.send('tree-view:project-folder-dropped', fromIndex)
+
+  removeDropTargetClasses: ->
+    workspaceElement = $(atom.views.getView(atom.workspace))
+    workspaceElement.find('.tree-view .is-drop-target').removeClass 'is-drop-target'
+    workspaceElement.find('.tree-view .drop-target-is-after').removeClass 'drop-target-is-after'
+
+  getDropTargetIndex: (event) ->
+    target = $(event.target)
+
+    return if @isPlaceholder(target)
+
+    projectRoots = $(@treeView.roots)
+    projectRoot = target.closest('.project-root')
+    projectRoot = projectRoots.last() if projectRoot.length is 0
+
+    return 0 unless projectRoot.length
+
+    element = projectRoot.find('.project-root-header')
+
+    elementCenter = element.offset().top + element.height() / 2
+
+    if event.originalEvent.pageY < elementCenter
+      projectRoots.index(projectRoot)
+    else if projectRoot.next('.project-root').length > 0
+      projectRoots.index(projectRoot.next('.project-root'))
+    else
+      projectRoots.index(projectRoot) + 1
+
+  canDragStart: (event) ->
+    $(event.target).closest('.project-root-header').size() > 0
+
+  isDragging: (event) ->
+    Boolean event.originalEvent.dataTransfer.getData 'from-root-path'
+
+  getPlaceholder: ->
+    @placeholderEl ?= $('<li/>', class: 'placeholder')
+
+  removePlaceholder: ->
+    @placeholderEl?.remove()
+    @placeholderEl = null
+
+  isPlaceholder: (element) ->
+    element.is('.placeholder')
+
+  getWindowId: ->
+    @processId ?= atom.getCurrentWindow().id
+
+  browserWindowForId: (id) ->
+    BrowserWindow ?= require('remote').require('browser-window')
```

Comment:
Unfortunately, i was stubbing that method, so I didn't catch this in the unit tests.
Good catch, thanks!